pytest_rerunfailures.py

```diff
@@ -122,6 +122,31 @@ def get_reruns_delay(item):
     return delay
 
 
+def _remove_cached_results_from_fixtures(item):
+    """
+    Note: remove all cached_result attribute from every fixture
+    """
+    cached_result = 'cached_result'
+    fixture_info = getattr(item, '_fixtureinfo')
+    for fixture_def_str in fixture_info.name2fixturedefs:
+        fixture_defs = fixture_info.name2fixturedefs[fixture_def_str]
+        for fixture_def in fixture_defs:
+            if hasattr(fixture_def, cached_result):
+                delattr(fixture_def, cached_result)
+
+
+def _remove_failed_setup_state_from_session(item):
```

Comment:
I am a little bit reluctant to introduce cleanup login in the plugin. This could lead to too much coupling. As the `SetupState` is not too complicated, it looks like https://github.com/pytest-dev/pytest/blob/master/src/_pytest/runner.py#L331 is actually doing the job already. Would that be a possibility?

Reply:
If I understand correctly, you advising to use SetupState.teardown_all() method, instead of a custom cleaner that I've implemented.

I've tried to use it, but it seems not handling the resource properly. All it does is calling all the `teadowns` for each `col` elements in the stack, but it is missing the most important part: invalidation/removing the failed execution result stored in `_prepare_exc` field. That's why every new test that is relying on `setup_class` will rerun it anyway, but it shouldn't because setup was already successfully re-executed after the first attempt.
You can see the problem by changing the body of `_remove_failed_setup_state_from_session` function to: `setup_state = getattr(item.session, '_setupstate') setup_state.teardown_all()` and running the test:
'''
class TestFoo(object):
    execution_number = 1
    srr = None

    @classmethod
    def setup_class(cls):
        cls.srr = Random.rand_char_string(10)
        Log.info('IN SETUP %s' % cls.srr)
        if cls.execution_number:
            cls.execution_number -= 1
            assert False
        assert True

    @mark.parametrize('param', [1, 2, 3])
    def test_pass(self, param):
        Log.info('IN TEST %s %s' % (param, self.srr))
        assert param
'''

The output will be:
'''
R2018-09-14 11:12:27,314 IN SETUP XWnfETJrmi
2018-09-14 11:12:27,568 IN SETUP ArdcndHZkT
.2018-09-14 11:12:27,569 IN TEST 1 ArdcndHZkT
R2018-09-14 11:12:27,617 IN SETUP GbnwDwvSWi
.2018-09-14 11:12:27,618 IN TEST 2 GbnwDwvSWi
R2018-09-14 11:12:27,642 IN SETUP tdWVHVPPyi
.2018-09-14 11:12:27,643 IN TEST 3 tdWVHVPPyi
'''

I do agree with you though, that instead of cleaning stack by myself, I should've used teardown_all().
With this change the output of the previous test is correct:
'''
R2018-09-14 11:13:36,147 IN SETUP biCSDlQwSP
2018-09-14 11:13:36,303 IN SETUP ZKlZjHlYDB
.2018-09-14 11:13:36,304 IN TEST 1 ZKlZjHlYDB
.2018-09-14 11:13:36,308 IN TEST 2 ZKlZjHlYDB
.2018-09-14 11:13:36,320 IN TEST 3 ZKlZjHlYDB
'''

Reply:
Further investigation showed that we cannot use `.teardown_all()` because it is triggering all the finalizers, even for the module/session scoped fixtures. Thus, leaving original implementation.

Reply:
Yeah, that is true. So I am happy with keeping the cleanup here. At this point a comment could be helpful.